test/unit-test/test_keystore_antiklepto.c

```diff
@@ -0,0 +1,183 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <setjmp.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <cmocka.h>
+
+#include <keystore.h>
+#include <keystore/keystore_antiklepto.h>
+
+#include <secp256k1_ecdsa_s2c.h>
+#include <wally_bip32.h>
+#include <wally_crypto.h>
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wnested-externs"
+
+static uint8_t _mock_seed[32] = {
+    0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22,
+    0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44,
+};
+
+static uint8_t _mock_bip39_seed[64] = {
+    0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22,
+    0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44,
+    0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x11, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22, 0x22,
+    0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44, 0x44,
+};
+
+bool __wrap_keystore_secp256k1_sign(
+    const uint32_t* keypath,
+    size_t keypath_len,
+    const uint8_t* msg32,
+    const uint8_t* host_nonce32,
+    uint8_t* sig_compact_out,
+    int* recid_out)
+{
+    check_expected(keypath);
+    check_expected(keypath_len);
+    check_expected(msg32);
+    check_expected(host_nonce32);
+    check_expected(sig_compact_out);
+    check_expected(recid_out);
+    return __real_keystore_secp256k1_sign(
+        keypath, keypath_len, msg32, host_nonce32, sig_compact_out, recid_out);
+}
+
+bool __wrap_keystore_secp256k1_nonce_commit(
+    const uint32_t* keypath,
+    size_t keypath_len,
+    const uint8_t* msg32,
+    const uint8_t* host_commitment,
+    uint8_t* signer_commitment_out)
+{
+    check_expected(keypath);
+    check_expected(keypath_len);
+    check_expected(msg32);
+    check_expected(host_commitment);
+    check_expected(signer_commitment_out);
+    return __real_keystore_secp256k1_nonce_commit(
+        keypath, keypath_len, msg32, host_commitment, signer_commitment_out);
+}
+
+static void _test_keystore_antiklepto(void** state)
+{
+    mock_state(_mock_seed, _mock_bip39_seed);
+
+    uint32_t keypath[] = {
+        84 + BIP32_INITIAL_HARDENED_CHILD,
+        1 + BIP32_INITIAL_HARDENED_CHILD,
+        0 + BIP32_INITIAL_HARDENED_CHILD,
+        0,
+        0,
+    };
+
+    uint8_t msg[32];
+    memset(msg, 0x23, sizeof(msg));
+
+    uint8_t host_nonce[32];
+    memset(host_nonce, 0x55, sizeof(host_nonce));
+
+    uint8_t signer_commitment[33] = {0};
+
+    uint8_t sig[64];
+    int recid;
+
+    // No cached data yet.
+    assert_false(keystore_antiklepto_secp256k1_sign(host_nonce, sig, &recid));
+
+    // Run multiple times to make sure the signing cache works a expected.
```

Comment:
I think making sure the cache is cleared on sign or abort is important enough to have a separate test case for it.

Reply:
Done